doc/man/man4/bpf.4.man
'\" rtp
.\" vim: ft=nroff sw=4 noet nocin nosi com=b\:.\\\" fo+=tcqlorn tw=77
.\" =========================================================================
.\"
.\" @(#) doc/man/man4/bpf.4.man
.\"
.\" =========================================================================
.\"
.\" Copyright (c) 2008-2011  Monavacon Limited <http://www.monavacon.com/>
.\" Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
.\" Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>
.\"
.\" All Rights Reserved.
.\"
.\" Permission is granted to copy, distribute and/or modify this manual
.\" under the terms of the GNU Free Documentation License, Version 1.3 or
.\" any later version published by the Free Software Foundation; with no
.\" Invariant Sections, no Front-Cover Texts, and no Back-Cover Texts.  A
.\" copy of the license is included in the section entitled "GNU Free
.\" Documentation License".
.\"
.\" Permission to use, copy and distribute this manual without modification,
.\" for any purpose and without fee or royalty is hereby granted, provided
.\" that both the above copyright notice and this permission notice appears
.\" in all copies and that the name of OpenSS7 Corporation not be used in
.\" advertising or publicity pertaning to distribution of this documentation
.\" or its contents without specific, written prior permission.  OpenSS7
.\" Corporation makes no representation about the suitability of this manual
.\" for any purpose.  It is provided "as is" without express or implied
.\" warranty.
.\"
.\" Permission is granted to process this file through groff and print the
.\" results, provided the printed document carries a copying permission
.\" notice identical to this one except for the removal of this paragraph
.\" (this paragraph not being relevant to the printed manual).
.\"
.\" OPENSS7 CORPORATION DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS MANUAL
.\" INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
.\" PARTICULAR PURPOSE, NON-INFRINGEMENT, OR TITLE; THAT THE CONTENTS OF THE
.\" DOCUMENT ARE SUITABLE FOR ANY PURPOSE, OR THAT THE IMPLEMENTATION OF
.\" SUCH CONTENTS WILL NOT INFRINGE ON ANY THIRD PARTY PATENTS, COPYRIGHTS,
.\" TRADEMARKS OR OTHER RIGHTS.  IN NO EVENT SHALL OPENSS7 CORPORATION BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, SPECIAL OR CONSEQUENTIAL DAMAGES OR ANY
.\" DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER
.\" IN AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING
.\" OUT OF OR IN CONNECTION WITH ANY USE OF THIS DOCUMENT OR THE PERFORMANCE
.\" OR IMPLEMENTATION OF THE CONTENTS THEREOF.
.\" 
.\" Since the Linux kernel and libraries are constantly changing, this
.\" manual page may be incorrect or out-of-date.  The author(s) assume no
.\" responsibility for errors or omissions, or for damages resulting from
.\" the use of the information contained herein.  The author(s) may not have
.\" taken the same level of care in the production of this manual, which is
.\" licensed free of charge, as they might when working professionally.  The
.\" author(s) will take no responsibility in it.
.\" 
.\" Formatted or processed versions of this manual, if unaccompanied by the
.\" source, must acknowledge the copyright and authors of this work.
.\"
.\" -------------------------------------------------------------------------
.\"
.\" U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software
.\" on behalf of the U.S. Government ("Government"), the following
.\" provisions apply to you.  If the Software is supplied by the Department
.\" of Defense ("DoD"), it is classified as "Commercial Computer Software"
.\" under paragraph 252.227-7014 of the DoD Supplement to the Federal
.\" Acquisition Regulations ("DFARS") (or any successor regulations) and the
.\" Government is acquiring only the license rights granted herein (the
.\" license rights customarily provided to non-Government users).  If the
.\" Software is supplied to any unit or agency of the Government other than
.\" DoD, it is classified as "Restricted Computer Software" and the
.\" Government's rights in the Software are defined in paragraph 52.227-19
.\" of the Federal Acquisition Regulations ("FAR") (or any successor
.\" regulations) or, in the cases of NASA, in paragraph 18.52.227-86 of the
.\" NASA Supplement to the FAR (or any successor regulations).
.\"
.\" =========================================================================
.\" 
.\" Commercial licensing and support of this software is available from
.\" OpenSS7 Corporation at a fee.  See http://www.openss7.com/
.\" 
.\" =========================================================================
.\"
.R1
bracket-label "\fR[\fB" "\fR]" "\fR,\fB "
no-default-database
database openss7.refs
accumulate
move-punctuation
abbreviate A
join-authors ", " ", " " and "
et-al " et al" 2 3
abbreviate-label-ranges ".."
sort-adjacent-labels
search-truncate 40
search-ignore CGIQOSTU
.R2
.so openss7.macros
.\"
.\"
.TH BPF 4 "@PACKAGE_DATE@" "@PACKAGE@-@VERSION@" "@PACKAGE_TITLE@ Modules"
.\"
.\"
.SH NAME
.B bpf
\-\fI STREAMS\fR Berkeley Packet Filter device
.\"
.\"
.SH SYNOPSIS
.PP
.B #include <stropts.h>
.br
.B #include <net/bpf.h>
.HP 8
.BI "int " fd ;
.PD 0
.HP
.BI "int open(" \*(lq\fR/dev/streams/bpf/\fP\fIxxxx\fP\*(rq ", " \fR0\fP );
.HP
.BI "int ioctl(" fd ", " I_PUSH ", " \*(lqbpfmod\*(rq );
.PD
.\"
.\"
.SH DESCRIPTION
.PP
The Berkeley Packet Filter provides a raw interface to data link layers in a
protocol independent fashion.  All packets on the network, even those destined
for other hosts, are accessible through this mechanism.
.PP
The packet filter appears as a character special device,
.BR /dev/bpf ,
.BR /dev/bpf\fIn\fP  " or " /dev/streams/bpf/\fIxxxx\fP ,
where
.I xxxx
is the name of a
.I STREAMS
driver supporting BPF.
.\"This special device is an alias for
.\".BR /dev/dlpi " or " /dev/streams/clone/dlpi
.\"that has an
.\".BR autopush (8)
.\"specification that includes this module,
.\".BR bpfmod .
Because the
.BR /dev/bpf ,
.BR /dev/bpf\fIn\fP " or " /dev/streams/bpf/\fIxxxx\fP
devices are clone devices, each time that an application opens the device, a
new character device special file is created just for that opener.
.PP
After opening the device, or pushing the module, the file descriptor must be
bound to a specific network interface with the
.I BIOCSETIF
command if it has not already been bound on the
stream before pushing the module.
A given interface can be shared with multiple
listeners, and the filter underlying each descriptor will see a identical
packet stream.
.PP
Associated with each open instance of a BPF file is a user-settable packet
filter.  The packet filter will support any link level protocol that has
fixed length headers.  Whenever a packet is received by an interface, all
file descriptors listening on that interface apply their filter.  Each
descriptor that accepts the packet receives it own copy.
.PP
Reads from these files return the next group of packets that have matched the
filter.  To improve performance, the buffer passed to read must be the same
size as the buffers used internally by BPF.  This size is returned by the
.I BIOCGBLEN
command (see below), and can be set with
.IR BIOCSBLEN .
Note that an individual packet larger than this size is necessarily
truncated.
.PP
A packet can be sent out on the network by writing to a BPF file descriptor.
The writes are unbuffered, meaning only one packet can be processed per
write.  Currently, only writes to Ethernets and SLIP links are supported.
.\"
.\"
.SH USAGE
.PP
The\fB libpcap\fP functions (part of\fB tcpdump\fP(1)) for BPF provide
excellent examples of the traditional usage of BPF across a very wide range
of operating systems and provides many insights into usage for different
implementations.  The steps taken by\fB libcap\fP to prepare a device and
activate capture are as follows:
.\"
.SS "Finding All Devices"
.PP
.B libpcap
finds as many devices as it can by opening a socket and using
the
.BR netdevice (7)
socket input-output control commands:
.BR SIOCGIFNUM " and " SIOCGIFCONF ,
or
.BR SIOCGLIFNUM " and " SIOCGLIFCONF ,
when available.  This lists, by name and transport address, all of the IPv4
(and IPv6 with the later pair) interface available.
.B libpcap
then tests each interface using the
.BR SIOCGIFFLAGS " or " SIOCGLIFFLAGS
commands to determine which interfaces advertise the
.I IFF_UP
flag indicating that the address is up.
The resulting list of interfaces is used as a list from which to select.
(For example,
.B wireshark
uses this list to display to the user a selection of interfaces.)
.\"
.\"
.SS "Initiating Capture"
.PP
.IP \ 1. \w'00.\(em'u
When the OS supports a cloning interface (such as AIX and some versions of
MAC), open the\fB /dev/bpf\fP device.  Otherwise, start opening devices
\fB/dev/bpf\fP\fIn\fP starting at\fB /dev/bpf\fP\fI0\fP until it is possible
to open a free device, or the device does not exist.
.IP ""
The lesson here is to provide both a\fB /dev/bpf\fP and\fB /dev/bpf\fP\fIn\fP
devices, where\fI n\fP is, say, from 0 to 8.  All these devices ultimately
link to (or have the device numbers of)\fB /dev/streams/bpf/ldl\fP.
.IP ""
The\fB @PACKAGE@\fP version of\fB libpcap\fP will also open devices
\fB/dev/streams/bpf/\fP\fIxxxx\fP, where\fI xxxx\fP is the name of a BPF
registered driver such as\fB x400p-ss7\fP.
.IP \ 2.
.IP \ 3.
.IP \ 4.
.IP \ 5.
.IP \ 6.
.IP \ 7.
.IP \ 8.
.IP \ 9.
.IP \10.
.IP \11.
.IP \12.
.IP \13.
.IP \14.
.IP \15.
.IP \16.
.PP
Use of the BPF device consists of the following steps:
.IP \ 1. \w'00.\(em'u
Select an appropriate device.
For an OpenSS7 SS7 signalling link, use\fI /dev/streams/bpf/x400p-ss7\fP.
\fBlibpcap\fP opens the\fB /dev/bpf\fP device if the system supports cloning
(AIX, MAC), otherwise, it starts walking down the list of
\fB/dev/bpf\fP\fIn\fP devices start at\fI n\fP=0 until it finds an
appropriate device.
.IP \ 2.
Open the device using the\fB open\fP(2s) system call.  It is not necessary to
push the\fB bpfmod\fP(4) module because an autopush specification will do
this as the device is opened.
When opening the device with the\fB O_RDWR\fP flag is denied [\fBEACCES\fP],
try opening the device with the\fB O_RDONLY\fP flag.
.IP \ 3.
Set the interface.   The interface can be set with the\fB BIOCSETIF\fP
command and subsequently examined using the
\fBBIOCGETIF\fP input output control command.  The [\fBENXIO\fP] or
[\fBENETDOWN\fP] errors are considered spearately by\fB libpcap\fP.
.IP \ 4.
\fBlibpcap\fP checks whether the\fB BIOCGDLTLIST\fP command is available and
can be used on the interface for later reference.  It checks whether any
wireless DLTs are available to determine whether RF monitoring modes are
available.
.IP \ 5.
Try to set up zero-copy buffer mode using the\fB BIOCSETBUFMODE\fP command.
If it fails, zero-copy buffer mode is not available.  If available, get the
maximum zero-copy buffer size using the\fB BIOCGETZMAX\fP command.  If no
buffer maximum specified or the request fails, use the minimum of the buffer
size returned by the\fB BIOCGBLEN\fP command and the value of
\fBBPF_DFLTBUFSIZE\fP.
.IP \ 6.
Check the BPF version.  The version is checked using the\fB BIOCVERSION\fP
command.  The default major version is\fB 1\R and the
device minor version is\fB 1\fP also.
.IP \ 7.
Set the buffer size.  The buffer size is checked using the\fB BIOCGBLEN\fP
command and can be set using the\fB BIOCSBLEN\fP
command.  The default size is\fB 1M\fP byte; the
minimum size\fB 32\fP, the maximum\fB 16M\fP.
.IP \ 8.
Set the read mode.  The read mode can be set using the\fB BIOCIMMEDIATE\fP
command.  The default setting is false (immediate
mode disabled).
.IP \ 9.
Set the read timeout.  The read timeout can be checked using the
\fBBIOCGRTIMEOUT\fP command and set using the
\fBBIOCSRTIMEOUT\fP command.  The default setting is
zero (0) which disables the read timeout.
.IP \11.
Set the filter program.  The filter program can be set using the
\fBBIOCSETF\fP command.  Setting a program with a
program length of zero (0) disables the program.  The default filter
disposition for a disabled program is to pass all packets.
.IP \12.
Set the data link type.  The data link type cannot be set  or examined until
after the interface is set.  One way to avoid unwanted traffic flowing is to
set a filter program that discards all messages.  The data link types
available for the specified interface can be checked using the
\fBBIOCGDLTLIST\fP command, set using the\fB BIOCSDLT\fP
command, and examined using the\fB BIOCGDLT\fP
command.
.IP \13.
Set promiscuous mode.  If promiscuous mode capture is desired, set
promiscuous mode on the interface using the\fB BIOCPROMISC\fP 
command.  Note that the command may return
[\fBEOPNOTSUPP\fP] or [\fBEINVAL\fP] when the interface does not support the
concept of a promiscuous mode.
.IP \14.

NEWS
OpenSS7 -- history of user visible changes.  2015-01-15
$Id: NEWS,v 1.1.2.25 2011-05-10 13:45:30 brian Exp $
Copyright (c) 2008-2015  Monavacon Limited. <http://www.monavacon.com/>
Copyright (c) 2001-2008  OpenSS7 Corporation. <http://www.openss7.com/>
Copyright (c) 1997-2001  Brian Bidulock <bidulock@openss7.org>
See the end for copying conditions (for this file).

The sections that follow provide information on OpenSS7 releases of the
'OpenSS7' package.

Major changes for release openss7-1.1.7.20150115
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Source code release on GitHub.

Major changes for release openss7-1.1.7.20131209
------------------------------------------------

This is the seventh release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build corrections for RHEL 6.4.

   - Scripts for country boundary, shoreline and rivers database
     preparation.

   - Database for ANSI T1.101 point code assignments.

   - Suppress extransous /proc/1/comm error messages when testing for
     systemd.

   - Handle PUIAS to Springdale rename for PUIAS distribution.

Major changes for release openss7-1.1.7.20131123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Correction to ca-cert handling in RHEL 6 install scriptlet.

   - RHEL 6 build changes.

   - Set permissions correctly when applying kernel module patch.

   - Do not duplicate certs when including ca-cert.

   - Avoid conflicting shelll variable in RPM install scriplet.

   - Remove /proc/1/comm checks for systemd in module loader.

Major changes for release openss7-1.1.7.20130209
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Corrections to RPM spec file.

Major changes for release openss7-1.1.7.20130129
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Compensate for RHEL6 lack of c_rehash.

   - Do not create sysfs nodes: they do not work properly.

   - Use no foreground flag in strace.service file.

   - More error traces for strerr logger.

   - Remove static device names (the do not work either).

   - Changes to support weak kernel modules.

   - Clean up weak module support.

Major changes for release openss7-1.1.7.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Better Archlinux install script support.

Major changes for release openss7-1.1.6.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Remove use of kill_litter_super().

Major changes for release openss7-1.1.3.20130125
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Updates and corrections to weak kernel module builds.

   - Repo support and corrections for Archlinux packages.

Major changes for release openss7-1.1.3.20130123
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Additional support for Archlinux.

   - Add udev rules for strace and strerr.

   - Add system unit files for strace and strerr.

   - Add repo support for Archlinux packages.

   - Build Archlinux packages better.

Major changes for release openss7-1.1.3.20130121
------------------------------------------------

This is the sixth release of the 'OpenSS7' package.  This is a
production grade release.  All existing validation test suites run clean
on supported distributions, production kernels and architectures.

Major features since the last release are as follows:

   - Build support for SLES 11.

   - Handle certificates better.

   - Get configuration files in the right places for SLES 11.

   - Support in RPM spec file for certificates on SLES 11.

   - Create a new OpenSS7 certificate.

   - Repository support for SLES 11.

   - RPM spec file reworked for new RPM release.

   - Better certificate handling for RHEL.

   - Additional build support for RHEL 6 with certificates.


src/drivers/bpf.c
/*****************************************************************************

 @(#) File: src/drivers/bpf.c

 -----------------------------------------------------------------------------

 Copyright (c) 2008-2015  Monavacon Limited <http://www.monavacon.com/>
 Copyright (c) 2001-2008  OpenSS7 Corporation <http://www.openss7.com/>
 Copyright (c) 1997-2001  Brian F. G. Bidulock <bidulock@openss7.org>

 All Rights Reserved.

 This program is free software: you can redistribute it and/or modify it under
 the terms of the GNU Affero General Public License as published by the Free
 Software Foundation, version 3 of the license.

 This program is distributed in the hope that it will be useful, but WITHOUT
 ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 FOR A PARTICULAR PURPOSE.  See the GNU Affero General Public License for more
 details.

 You should have received a copy of the GNU Affero General Public License
 along with this program.  If not, see <http://www.gnu.org/licenses/>, or
 write to the Free Software Foundation, Inc., 675 Mass Ave, Cambridge, MA
 02139, USA.

 -----------------------------------------------------------------------------

 U.S. GOVERNMENT RESTRICTED RIGHTS.  If you are licensing this Software on
 behalf of the U.S. Government ("Government"), the following provisions apply
 to you.  If the Software is supplied by the Department of Defense ("DoD"), it
 is classified as "Commercial Computer Software" under paragraph 252.227-7014
 of the DoD Supplement to the Federal Acquisition Regulations ("DFARS") (or any
 successor regulations) and the Government is acquiring only the license rights
 granted herein (the license rights customarily provided to non-Government
 users).  If the Software is supplied to any unit or agency of the Government
 other than DoD, it is classified as "Restricted Computer Software" and the
 Government's rights in the Software are defined in paragraph 52.227-19 of the
 Federal Acquisition Regulations ("FAR") (or any successor regulations) or, in
 the cases of NASA, in paragraph 18.52.227-86 of the NASA Supplement to the FAR
 (or any successor regulations).

 -----------------------------------------------------------------------------

 Commercial licensing and support of this software is available from OpenSS7
 Corporation at a fee.  See http://www.openss7.com/

 *****************************************************************************/

static char const ident[] = "src/drivers/bpf.c (" PACKAGE_ENVR ") " PACKAGE_DATE;

/*
 * This is a bpf driver that works like the clone driver.  It accepts registration from other
 * drivers and provides a /dev/streams/bpf major device that has a /dev/streams/bpf/[driver] minor
 * device for each driver that registers.  [driver] is the driver name.  When the device is opened,
 * instead of forwarding to (minor,0) as does the clone driver, it forwards to (minor,2) with the
 * CLONEOPEN flag set.  There is also a dlpi and bpf driver that does a similar thing.
 */

#ifdef NEED_LINUX_AUTOCONF_H
#include NEED_LINUX_AUTOCONF_H
#endif
#include <linux/version.h>
#include <linux/module.h>
#include <linux/init.h>
#include <linux/sched.h>

#include <sys/stream.h>
#include <sys/strconf.h>
#include <sys/strsubr.h>
#include <sys/ddi.h>

#include <net/bpf.h>		/* extern verification */

#include "sys/config.h"

#define BPF_DESCRIP	"Berkeley Packet Filter (BPF) STREAMS Driver"
#define BPF_EXTRA	"Part of UNIX SYSTEM V RELEASE 4.2 FAST STREAMS FOR LINUX"
#define BPF_COPYRIGHT	"Copyright (c) 2008-2015  Monavacon Limited.  All Rights Reserved."
#define BPF_REVISION	"OpenSS7 src/drivers/bpf.c (" PACKAGE_ENVR ") " PACKAGE_DATE
#define BPF_DEVICE	"SVR 4.2 MP STREAMS BPF Driver"
#define BPF_CONTACT	"Brian Bidulock <bidulock@openss7.org>"
#define BPF_LICENSE	"GPL"
#define BPF_BANNER	BPF_DESCRIP	"\n" \
			BPF_EXTRA	"\n" \
			BPF_COPYRIGHT	"\n" \
			BPF_REVISION	"\n" \
			BPF_DEVICE	"\n" \
			BPF_CONTACT	"\n"
#define BPF_SPLASH	BPF_DEVICE	" - " \
			BPF_REVISION	"\n"

#ifdef CONFIG_STREAMS_BPF_MODULE
MODULE_AUTHOR(BPF_CONTACT);
MODULE_DESCRIPTION(BPF_DESCRIP);
MODULE_SUPPORTED_DEVICE(BPF_DEVICE);
MODULE_LICENSE(BPF_LICENSE);
#ifdef MODULE_VERSION
MODULE_VERSION(PACKAGE_ENVR);
#endif				/* MODULE_VERSION */
#endif				/* CONFIG_STREAMS_BPF_MODULE */

#ifndef CONFIG_STREAMS_BPF_NAME
#error "CONFIG_STREAMS_BPF_NAME must be defined."
#endif
#ifndef CONFIG_STREAMS_BPF_MODID
#error "CONFIG_STREAMS_BPF_MODID must be defined."
#endif
#ifndef CONFIG_STREAMS_BPF_MAJOR
#error "CONFIG_STREAMS_BPF_MAJOR must be defined."
#endif

#ifdef CONFIG_STREAMS_BPF_MODULE
modID_t bpf_modid = CONFIG_STREAMS_BPF_MODID;

#ifndef module_param
MODULE_PARM(bpf_modid, "h");
#else
module_param(bpf_modid, ushort, 0444);
#endif
MODULE_PARM_DESC(bpf_modid, "Module id number for BPF driver.");
#else				/* CONFIG_STREAMS_BPF_MODULE */
static modID_t bpf_modid = CONFIG_STREAMS_BPF_MODID;
#endif				/* CONFIG_STREAMS_BPF_MODULE */

#ifdef CONFIG_STREAMS_BPF_MODULE
major_t major = CONFIG_STREAMS_BPF_MAJOR;

#ifndef module_param
MODULE_PARM(major, "h");
#else
module_param(major, uint, 0444);
#endif
MODULE_PARM_DESC(major, "Major device number for BPF driver.");
#else				/* CONFIG_STREAMS_BPF_MODULE */
static major_t major = CONFIG_STREAMS_BPF_MAJOR;
#endif				/* CONFIG_STREAMS_BPF_MODULE */

#ifdef MODULE
#ifdef MODULE_ALIAS
MODULE_ALIAS("streams-bpf");
MODULE_ALIAS("streams-modid-" __stringify(CONFIG_STREAMS_BPF_MODID));
MODULE_ALIAS("streams-driver-bpf");
MODULE_ALIAS("char-major-" __stringify(CONFIG_STREAMS_BPF_MAJOR) "-*");
MODULE_ALIAS("streams-major-" __stringify(CONFIG_STREAMS_BPF_MAJOR));
MODULE_ALIAS("/dev/streams/bpf");
MODULE_ALIAS("/dev/streams/bpf/*");
#endif
#endif

static struct module_info bpf_minfo = {
	.mi_idnum = CONFIG_STREAMS_BPF_MODID,
	.mi_idname = CONFIG_STREAMS_BPF_NAME,
	.mi_minpsz = STRMINPSZ,
	.mi_maxpsz = STRMAXPSZ,
	.mi_hiwat = STRHIGH,
	.mi_lowat = STRLOW,
};

static struct module_stat bpf_rstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));
static struct module_stat bpf_wstat __attribute__ ((__aligned__(SMP_CACHE_BYTES)));

static struct qinit bpf_rinit = {
	.qi_putp = strrput,
	.qi_srvp = strrsrv,
	.qi_qopen = str_open,
	.qi_qclose = str_close,
	.qi_minfo = &bpf_minfo,
	.qi_mstat = &bpf_rstat,
};

static struct qinit bpf_winit = {
	.qi_putp = strwput,
	.qi_srvp = strwsrv,
	.qi_minfo = &bpf_minfo,
	.qi_mstat = &bpf_wstat,
};

#ifdef CONFIG_STREAMS_BPF_MODULE
static
#endif
struct streamtab bpfinfo = {
	.st_rdinit = &bpf_rinit,
	.st_wrinit = &bpf_winit,
};

/**
 *  bpfopen: - open a bpf special device
 *  @inode:	shadow special filesystem inode
 *  @file:	shadow special filesystem file pointer
 *
 *  bpfopen() is called only from within the shadow special filesystem.  This can occur by
 *  chaining from the external filesystem (e.g. openining a character device with bpf major) or by
 *  direct open of the inode within the mounted shadow special filesystem.  Either way, the inode
 *  number has our extended device numbering as a inode number and we chain the call within the
 *  shadow special filesystem.
 */
static int
bpfopen(struct inode *inode, struct file *file)
{
	struct cdevsw *cdev;
	dev_t dev = inode->i_ino;

	if (file->private_data)
		/* Darn.  Somebody passed us a FIFO inode. */
		return (-EIO);

	if ((cdev = sdev_get(getminor(dev)))) {
		int err;

		dev = makedevice(cdev->d_modid, BPF_NODE);
		err = spec_open(file, cdev, dev, CLONEOPEN);
		sdev_put(cdev);
		return (err);
	}
	return (-ENOENT);
}

struct file_operations bpf_ops ____cacheline_aligned = {
	.owner = THIS_MODULE,
	.open = bpfopen,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

static struct cdevsw bpf_cdev = {
	.d_name = "bpf",
	.d_str = &bpfinfo,
	.d_flag = D_CLONE | D_MP,
	.d_fop = &bpf_ops,
	.d_mode = S_IFCHR | S_IRUGO | S_IWUGO,
	.d_kmod = THIS_MODULE,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  REGISTRATION
 *
 *  -------------------------------------------------------------------------
 */

streams_fastcall int
register_bpf(struct cdevsw *cdev)
{
	int err;
	struct devnode *cmin, *node;
	struct strapush *sap;

	err = -ENOMEM;
	if (!(cmin = kmalloc(2 * sizeof(*cmin) + sizeof(*sap), GFP_ATOMIC))) {
		goto error;
	}
	memset(cmin, 0, sizeof(*cmin));
	INIT_LIST_HEAD(&cmin->n_list);
	INIT_LIST_HEAD(&cmin->n_hash);
	cmin->n_name = cdev->d_name;
	cmin->n_str = cdev->d_str;
	cmin->n_flag = bpf_cdev.d_flag;
	cmin->n_modid = cdev->d_modid;
	cmin->n_count = (atomic_t) ATOMIC_INIT(0);
	cmin->n_sqlvl = cdev->d_sqlvl;
	cmin->n_syncq = cdev->d_syncq;
	cmin->n_kmod = cdev->d_kmod;
	cmin->n_major = bpf_cdev.d_major;
	cmin->n_mode = bpf_cdev.d_mode;
	cmin->n_minor = cdev->d_major;
	cmin->n_dev = &bpf_cdev;
	if ((err = register_strnod(&bpf_cdev, cmin, cdev->d_major)) < 0) {
		kfree(cmin);
		goto error;
	}
	node = cmin + 1;
	memset(node, 0, sizeof(*node));
	INIT_LIST_HEAD(&node->n_list);
	INIT_LIST_HEAD(&node->n_hash);
	node->n_name = "bpf";
	node->n_str = cdev->d_str;
	node->n_flag = D_CLONE;
	node->n_modid = cdev->d_modid;
	node->n_count = (atomic_t) ATOMIC_INIT(0);
	node->n_sqlvl = cdev->d_sqlvl;
	node->n_syncq = cdev->d_syncq;
	node->n_kmod = cdev->d_kmod;
	node->n_major = cdev->d_major;
	node->n_mode = cdev->d_mode;
	node->n_minor = BPF_NODE;
	node->n_dev = cdev;
	if ((err = register_strnod(cdev, node, BPF_NODE)) < 0) {
		unregister_strnod(&bpf_cdev, cdev->d_major);
		kfree(cmin);
		goto error;
	}
	sap = (typeof(sap)) (node + 1);
	memset(sap, 0, sizeof(*sap));
	sap->sap_cmd = SAP_CLONE;
	sap->sap_major = cdev->d_major;
	sap->sap_minor = BPF_NODE;
	sap->sap_lastminor = -1;
	sap->sap_npush = 1;
	snprintf(sap->sap_list[0], FMNAMESZ+1, "bfpmod");
	if ((err = autopush_add(sap)) < 0) {
		unregister_strnod(cdev, BPF_NODE);
		unregister_strnod(&bpf_cdev, cdev->d_major);
		kfree(cmin);
		goto error;
	}
      error:
	return (err);
}

EXPORT_SYMBOL_GPL(register_bpf);

streams_fastcall int
unregister_bpf(struct cdevsw *cdev)
{
	int err;
	struct devnode *cmin;
	struct strapush *sap;

	err = -ENXIO;
	if (!(cmin = cmin_get(&bpf_cdev, cdev->d_major)))
		goto error;
	sap = (typeof(sap)) (cmin + 2);
	if ((err = autopush_del(sap)) < 0)
		goto error;
	if ((err = unregister_strnod(cdev, BPF_NODE)))
		goto error;
	if ((err = unregister_strnod(&bpf_cdev, cdev->d_major)))
		goto error;
	kfree(cmin);
      error:
	return (err);
}

EXPORT_SYMBOL_GPL(unregister_bpf);

/* 
 *  -------------------------------------------------------------------------
 *
 *  Special open for bpf devices.
 *
 *  -------------------------------------------------------------------------
 */

/*
 *  bpf_open: - open a bpf device node
 *  @inode: the external filesystem inode
 *  @file: the external filesystem file pointer
 *
 *  bpf_open() is only used to open a bpf device from a character device node in an external
 *  filesystem.  This is never called for direct opens of a specfs device node (for direct opens see
 *  spec_dev_open() in strspecfs.c).  The character device number from the inode is used to
 *  determine the shadow special filesystem (internal) inode and chain the open call.
 *
 *  This is the separation point where we convert the external device number to an internal device
 *  number.  The external device number is contained in inode->i_rdev.
 */
STATIC int
bpf_open(struct inode *inode, struct file *file)
{
	int err;
	struct cdevsw *cdev;
	major_t major;
	minor_t minor;
	modID_t modid, instance;

#if defined HAVE_KFUNC_TO_KDEV_T
	minor = MINOR(kdev_t_to_nr(inode->i_rdev));
	major = MAJOR(kdev_t_to_nr(inode->i_rdev));
#else
	minor = MINOR(inode->i_rdev);
	major = MAJOR(inode->i_rdev);
#endif
	minor = cdev_minor(&bpf_cdev, major, minor);
	major = bpf_cdev.d_major;
	modid = bpf_cdev.d_modid;
	err = -ENXIO;
	if (!(cdev = sdev_get(minor))) {
		goto exit;
	}
	instance = cdev->d_modid;
	err = spec_open(file, cdev, makedevice(modid, instance), CLONEOPEN);
	sdev_put(cdev);
      exit:
	return (err);
}

STATIC struct file_operations bpf_f_ops ____cacheline_aligned = {
	.owner = NULL,		/* yes NULL */
	.open = bpf_open,
};

/* 
 *  -------------------------------------------------------------------------
 *
 *  INITIALIZATION
 *
 *  -------------------------------------------------------------------------
 */

#ifdef CONFIG_STREAMS_BPF_MODULE
static
#endif
int __init
bpfinit(void)
{
	int err;

#ifdef CONFIG_STREAMS_BPF_MODULE
	printk(KERN_INFO BPF_BANNER);
#else
	printk(KERN_INFO BPF_SPLASH);
#endif
	bpf_minfo.mi_idnum = bpf_modid;
	if ((err = register_cmajor(&bpf_cdev, major, &bpf_f_ops)) < 0)
		return (err);
	if (major == 0 && err > 0)
		major = err;
	return (0);
};

#ifdef CONFIG_STREAMS_BPF_MODULE
static
#endif
void __exit
bpfexit(void)
{
	if (unregister_cmajor(&bpf_cdev, major) != 0)
		swerr();
};

#ifdef CONFIG_STREAMS_BPF_MODULE
module_init(bpfinit);
module_exit(bpfexit);
#endif
